### DISCUSSION TITLE
How to make a workflow file to save the output as a file to my GitHub Repo?

### DISCUSSION THREAD

**Discussion by KeshariPiyush24** [community/community#26672] [Actions]
<p><strong>For past 3 days I have been learning about matplotlib and pandas from youtube and I tried to implement my knowledge by making a very basic python project,which analyzes data from an api hosted by the Govt of India containing the data of COVID-19 cases in different states of India.</strong></p>
<p><strong>My project is almost complete and is working fine in my local machine and I have hosted it to github pages as well.</strong></p>
<p><strong>The problem is that I wrote a python script to convert jupyter notebook to static <em>index.html</em> in the root folder and it works fine on my local machine.Basically my jupyter notebook takes data from the API source and analyzes it to display useful information,but GitHub pages cannot host jupyter notebooks so the script converts notebook to <em>index.html</em> file and saves it directly to root directory (Idea is to make index.html serve as the home page for GitHub pages) but when I made a workflow file to serve the purpose and when it executed something unexpected happened.</strong></p>
<p><strong>Below is My Workflow file <em>run.yml located in .github/workflows/run.yml</em></strong></p>
<pre><code>name: Update index.html

on:
  schedule:
    - cron: '* * * * *'

jobs:
  run:
    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python
      uses: actions/setup-python@v1
      with:
        python-version: '3.x'
    - name: Install dependencies
      run: |
        dir
        python -m pip install --upgrade pip
        pip install -r requirements.txt
        python update.py
        dir
</code></pre>
<p><strong><em>NOTE: I have Used dir commands to see where the index.html file is getting stored.</em></strong></p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4c51fc5e77501c5b95f50889662f895f6c8d5203.png" data-download-href="/uploads/short-url/aT9Wnl5XWz3Dw4pYw5wZk7fZHrB.png?dl=1" title="one.PNG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/18307180/181099134-b2c28820-6ab3-40e6-a712-3a29ce8fbd6e.png" alt="one.PNG" data-base62-sha1="aT9Wnl5XWz3Dw4pYw5wZk7fZHrB" width="690" height="246" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4c51fc5e77501c5b95f50889662f895f6c8d5203_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">one.PNG</span><span class="informations">959×342 12.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div> <strong>Output of first dir command</strong>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4094012724c15dfa4e517897bc4a7eed99cfd4fb.png" data-download-href="/uploads/short-url/9dhEGNvbU7K1bFLFYKmXqwPt2Yz.png?dl=1" title="two.PNG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/18307180/181099135-eeb4c185-49e7-4bbc-aa34-6c2ab9d1173b.png" alt="two.PNG" data-base62-sha1="9dhEGNvbU7K1bFLFYKmXqwPt2Yz" width="690" height="369" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4094012724c15dfa4e517897bc4a7eed99cfd4fb_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">two.PNG</span><span class="informations">746×400 25.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div> <strong>Output of second dir command</strong>
<p><strong>Seems like my script (update.py) is working fine but instead of saving <em>index.html</em> to the root directory of my Github Repo it is saving it somewhere else. I need help to make changes in my code to make index.html file to be saved on my GitHub repo as it will serve as the home page for my Github Pages.</strong></p>
<p><strong><a href="https://github.com/PiyushKe/COVID-19-Outbreak-India" rel="nofollow noopener">GitHub Repo Link</a></strong></p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Hi <a class="mention" href="/piyushke">@piyushke</a> , </p>
<p>Based on my understanding of your requirement, you want to push the index.html which created/changed during the workflow execution to your GitHub Repo. </p>
<p>You could add some steps to <strong>commit changes</strong> and <strong>push</strong> back to current repo. There is my example workflow yml:</p>
<pre><code>name: Update index.html

on: push

jobs:
  run:
    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python
      uses: actions/setup-python@v1
      with:
        python-version: '3.x'
    - name: Install dependencies
      run: |
        dir
        python -m pip install --upgrade pip
        pip install -r requirements.txt
        python update.py
        dir

    - name: Commit files
      id: commit
      run: |
        git config --local user.email "action@github.com"
        git config --local user.name "github-actions"
        git add --all
        if [-z "$(git status --porcelain)"]; then
           echo "::set-output name=push::false"
        else
           git commit -m "Add changes" -a
           echo "::set-output name=push::true"
        fi
      shell: bash
    - name: Push changes
      if: steps.commit.outputs.push == 'true'
      uses: ad-m/github-push-action@master
      with:
         github_token: ${{ secrets.GITHUB_TOKEN }}
</code></pre>
<p> If I misunderstood your scenario, please feel free to point me out. </p>

**Comment by PhilipeRLeal** [comment]
<p>Dear yanjingzhu, thank you for your reply.</p>
<p>While studying your answer, I ended up with 2 questions:</p>
<ol>
<li>
<p>how did you deal with infinite cycle of commiting and pushing into the remote repository? From what I understood, the moment the first push to this GitHub repository happens, it’s workflow will impose in a new push into itself, entering in an infinite loop of “git push origin master”.</p>
</li>
<li>
<p>what is the intent of line “if: if: steps.commit.outputs.push == 'true == 'true”? Where does one define this variable “steps.commit.outputs.push” in the workflow? Is it some hidden variable automatically defined by the Github’s action?</p>
</li>
</ol>

**Reply by Pamellamartins** [reply]
I guess you could set a trigger to only execute the workflow that has a tag on push, this would solve the problem with the loop on push.

**Reply by airtower-luna** [reply]
Using `GITHUB_TOKEN` to authorize the push [prevents such loops implicitly](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow):

> When you use the repository's GITHUB_TOKEN to perform tasks, events triggered by the GITHUB_TOKEN, with the exception of workflow_dispatch and repository_dispatch, will not create a new workflow run.
